hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1667,7 +1688,11 @@ protected HFileBlock readBlockDataInternal(FSDataInputStream is, long offset,
       final AttributesBuilder attributesBuilder = Attributes.builder();
       Optional.of(Context.current()).map(val -> val.get(CONTEXT_KEY))
         .ifPresent(c -> c.accept(attributesBuilder));
-      int onDiskSizeWithHeader = checkAndGetSizeAsInt(onDiskSizeWithHeaderL, hdrSize);
+      if (!checkOnDiskSizeWithHeader(onDiskSizeWithHeaderL)) {
```

Comment:
This code is definitely an improvement, but I still find it a bit hard to follow. It seems like we want to do the following:

1. safely convert the passed value from long to int
2. if the value is -1, try to read it from the header
3. if there is no cached header, fetch that first
4. finally we have a onDiskSizeWithHeader through one means or another, and now we should validate it

Any chance we can lay the code out like that, rather than do validation in multiple places? Something like this:

```
int onDiskSizeWithHeader = checkAndGetSizeAsInt(onDiskSizeWithHeaderL); // throw exception if value is greater than int.max
ByteBuff headerBuf = getCachedHeader(offset);

// make sure we have a size
if (onDiskSizeWithHeader == -1) {
  if (headerBuf == null) {
    // allocate and read headerBuf as today
  }
  onDiskSizeWithHeader = getOnDiskSizeWithHeader(headerBuf, checksumSupport);
}

// now we have a size, validate it
if (!checkOnDiskSizeWithHeader(onDiskSizeWithHeader)) {
  if (verifyChecksum) {
    invalidateNextBlockHeader();
    span.addEvent("Falling back to HDFS checksumming.", attributesBuilder.build());
    return null;
  } else {
    throw new IOException("Read invalid onDiskSizeWithHeader=" + onDiskSizeWithHeader);
  }
}
```

Reply:
That's fair, but also I think it might be more error prone (both to code and review) to try to bandaid it on since we all have to keep a few different code paths in mind. Like how we need to add another call to checkOnDiskSizeWithHeader for the cached header case.  It might seem slightly more invasive to do a reorganization of the code path in this method, but it also seems the least risky option imo. Hopefully it can backport cleanly since this file is not a hotspot for modifications. Looking at the git history, I think your tracing/span stuff might be biggest blocker for 2.4